test/VersionUpdater/SelfUpdateTest.php

```diff
@@ -0,0 +1,254 @@
+    }
+
+    public function testFailWhenDownloadedArchiveIsNotValid()
+    {
+        $installer = $this->getMockInstaller(['isValidSource']);
+        $installer
+            ->method('isValidSource')
+            ->willReturn(false);
+        $checker = $this->getMockChecker();
+        $output = $this->getMockOutput('not a valid archive');
+
+        $selfUpdate = $this->getSelfUpdater($checker, $installer);
+        $returnValue = $selfUpdate->run($this->getInput(), $output);
+
+        $this->assertEquals(SelfUpdate::FAILURE, $returnValue);
+    }
+
+    public function testFailWhenCreateBackupFails()
+    {
+        $installer = $this->getMockInstaller(['createBackup']);
+        $installer
+            ->method('createBackup')
+            ->willReturn(false);
+        $checker = $this->getMockChecker();
+        $output = $this->getMockOutput('Failed to create a backup');
+
+        $selfUpdate = $this->getSelfUpdater($checker, $installer);
+        $returnValue = $selfUpdate->run($this->getInput(), $output);
+
+        $this->assertEquals(SelfUpdate::FAILURE, $returnValue);
+    }
+
+    public function testBackupIsRestoredWhenInstallFails()
+    {
+        $installer = $this->getMockInstaller(['install', 'restoreFromBackup']);
+        $installer
+            ->method('install')
+            ->willReturn(false);
+        $installer
+            ->expects($this->once())
+            ->method('restoreFromBackup');
+        $checker = $this->getMockChecker();
+        $output = $this->getMockOutput();
+
+        $selfUpdate = $this->getSelfUpdater($checker, $installer);
+        $returnValue = $selfUpdate->run($this->getInput(), $output);
+
+        $this->assertEquals(SelfUpdate::FAILURE, $returnValue);
+    }
+
+    public function testExceptionIsNotCaughtWhenRestoreFails()
+    {
+        $this->expectException(ErrorException::class);
+        $installer = $this->getMockInstaller(['restoreFromBackup', 'install']);
+        $installer
+            ->method('install')
+            ->willReturn(false);
+        $installer
+            ->method('restoreFromBackup')
+            ->willThrowException(new ErrorException('Uncaught Exception'));
+
+        $checker = $this->getMockChecker();
+        $output = $this->getMockOutput();
+
+        $selfUpdate = $this->getSelfUpdater($checker, $installer);
+        $selfUpdate->run($this->getInput(), $output);
+
+        $this->fail('Expected ErrorException not thrown');
+    }
+
+    private function getInput()
+    {
+        $input = new ArgvInput([]);
+        // build a simple input with options that are used in SelfUpdate
+        $input->bind(new InputDefinition([
+            new InputOption('verbose', 'v', InputOption::VALUE_NONE)
+        ]));
+
+        return $input;
+    }
+
+    private function getMockChecker(bool $isLatest = false, string $version = Shell::VERSION)
+    {
+        $checker = $this->getMockBuilder(Checker::class)
+            ->onlyMethods(['getLatest', 'isLatest'])
```

Comment:
`onlyMethods` isn't a thing on PHPUnit versions old enough to run on 7.0 and 7.1. Honestly, I'd just skip these tests on PHP <7.2 if it's a pain to do another way.